### ISSUE TITLE
Gradient overflow on distributed training. Training stuck.

### ISSUE THREAD

**Issue by dimeldo** [bme-chatbots/dialogue-generation#10]
Machine: AWS provided 8 Nvidia V100 GPU's.
Environment: https://docs.nvidia.com/deeplearning/frameworks/tensorflow-release-notes/rel_19.02.html
```
Selected optimization level O2:  FP16 training with FP32 batchnorm and FP32 master weights.

Defaults for this optimization level are:
keep_batchnorm_fp32    : True
loss_scale             : dynamic
patch_torch_functions  : False
cast_model_type        : torch.float16
enabled                : True
master_weights         : True
opt_level              : O2
Processing user overrides (additional kwargs that are not None)...
After processing overrides, optimization options are:
keep_batchnorm_fp32    : True
loss_scale             : dynamic
patch_torch_functions  : False
cast_model_type        : torch.float16
enabled                : True
master_weights         : True
opt_level              : O2
2019-08-29 11:43:18,309 - INFO - {'max_epochs': 15, 'model_dir': '/Downloads/dialogue-generation/src/../model', 'top_k': 100, 'batch_size': 64, 'patience': 5, 'download_dir': '/Downloads/dialogue-generation/src/../data', 'max_hist': 4, 'max_len': 100, 'eval_every_step': 3000, 'grad_accum_steps': 2, 'model_name': 'gpt2-medium', 'mode': 'train', 'cuda': True, 'data_name': 'dailydialog', 'master_addr': '127.0.0.1', 'distributed': True, 'method': 'nucleus', 'master_port': '29500', 'num_devices': 8, 'top_p': 0.9, 'file_size': 100000, 'mixed': True, 'data_dir': '/Downloads/dialogue-generation/src/../data', 'learning_rate': 0.0001}
Train 0:   0%|                                                                                              | 0/468 [00:00<?, ?it/s]/usr/local/lib/python3.5/dist-packages/torch/distributed/distributed_c10d.py:100: UserWarning: torch.distributed.reduce_op is deprecated, please use torch.distributed.ReduceOp instead
  warnings.warn("torch.distributed.reduce_op is deprecated, please use "
/usr/local/lib/python3.5/dist-packages/torch/distributed/distributed_c10d.py:100: UserWarning: torch.distributed.reduce_op is deprecated, please use torch.distributed.ReduceOp instead
  warnings.warn("torch.distributed.reduce_op is deprecated, please use "
/usr/local/lib/python3.5/dist-packages/torch/distributed/distributed_c10d.py:100: UserWarning: torch.distributed.reduce_op is deprecated, please use torch.distributed.ReduceOp instead
  warnings.warn("torch.distributed.reduce_op is deprecated, please use "
/usr/local/lib/python3.5/dist-packages/torch/distributed/distributed_c10d.py:100: UserWarning: torch.distributed.reduce_op is deprecated, please use torch.distributed.ReduceOp instead
  warnings.warn("torch.distributed.reduce_op is deprecated, please use "
/usr/local/lib/python3.5/dist-packages/torch/distributed/distributed_c10d.py:100: UserWarning: torch.distributed.reduce_op is deprecated, please use torch.distributed.ReduceOp instead
  warnings.warn("torch.distributed.reduce_op is deprecated, please use "
/usr/local/lib/python3.5/dist-packages/torch/distributed/distributed_c10d.py:100: UserWarning: torch.distributed.reduce_op is deprecated, please use torch.distributed.ReduceOp instead
  warnings.warn("torch.distributed.reduce_op is deprecated, please use "
/usr/local/lib/python3.5/dist-packages/torch/distributed/distributed_c10d.py:100: UserWarning: torch.distributed.reduce_op is deprecated, please use torch.distributed.ReduceOp instead
  warnings.warn("torch.distributed.reduce_op is deprecated, please use "
/usr/local/lib/python3.5/dist-packages/torch/distributed/distributed_c10d.py:100: UserWarning: torch.distributed.reduce_op is deprecated, please use torch.distributed.ReduceOp instead
  warnings.warn("torch.distributed.reduce_op is deprecated, please use "
Train 0:   0%|▏                                                           | 1/468 [00:06<52:48,  6.79s/it, acc=0, loss=39.7, skip=0]Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 16384.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 16384.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 16384.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 16384.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 16384.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 16384.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 16384.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 16384.0
Train 0:   1%|▍                                                           | 3/468 [00:07<28:02,  3.62s/it, acc=0, loss=40.4, skip=0]Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 4096.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 4096.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 4096.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 4096.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 4096.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 4096.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 4096.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 4096.0
Train 0:   1%|▋                                                           | 5/468 [00:09<15:54,  2.06s/it, acc=0, loss=40.1, skip=0]Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 1024.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 1024.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 1024.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 1024.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 1024.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 1024.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 1024.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 1024.0
Train 0:   1%|▉                                                           | 7/468 [00:10<09:58,  1.30s/it, acc=0, loss=39.8, skip=0]Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 256.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 256.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 256.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 256.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 256.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 256.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 256.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 256.0
Train 0:   2%|█▏                                                          | 9/468 [00:11<07:05,  1.08it/s, acc=0, loss=40.5, skip=0]Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 128.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 128.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 128.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 128.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 128.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 128.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 128.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 128.0
Train 0:   3%|█▋                                                         | 13/468 [00:13<05:02,  1.50it/s, acc=0, loss=26.2, skip=0]Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 64.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 64.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 64.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 64.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 64.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 64.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 64.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 64.0
Train 0:   3%|█▉                                                         | 15/468 [00:14<04:38,  1.63it/s, acc=0, loss=24.6, skip=0]
```

**Comment by dimeldo** [comment]
It also gets stuck when `mixed` is set to `False`.

UPDATE: Training seems to go fine if `num_devices` is set to 1.

**Comment by dimeldo** [comment]
Because I'm using gpt2-medium on a big dataset the batch size needs to be 2 in order not to get OOM issues and skipping all the steps. This is why it's necessary for me to use distributed training.

**Comment by dimeldo** [comment]
Still an active problem...

**Comment by Mrpatekful** [comment]
This should be working correctly now.
